services/updates-svc/internal/updates/grpc.go

```diff
@@ -0,0 +1,133 @@
+package updates
+
+import (
+	"common/auth"
+	"context"
+	pb "gen/services/updates_svc/v1"
+	"github.com/EventStore/EventStore-Client-Go/v4/esdb"
+	"hwes"
+	"telemetry"
+	"time"
+)
+
+type UpdatesGrpcService struct {
+	pb.UnimplementedUpdatesServiceServer
+	es *esdb.Client
+}
+
+func NewUpdatesGrpcService(es *esdb.Client) *UpdatesGrpcService {
+	return &UpdatesGrpcService{
+		es: es,
+	}
+}
+
+func (s *UpdatesGrpcService) ReceiveUpdates(req *pb.ReceiveUpdatesRequest, stream pb.UpdatesService_ReceiveUpdatesServer) error {
+	ctx, span, log := telemetry.StartSpan(stream.Context(), "updates.ReceiveUpdates")
+	defer span.End()
+
+	tokenExpires, err := auth.SessionValidUntil(ctx)
+	if err != nil {
+		return err
+	}
+
+	tokenExpiresIn := time.Until(tokenExpires)
+
+	ctx, cancel := context.WithCancel(ctx)
+	// Close stream after session is invalid
+	time.AfterFunc(tokenExpiresIn, func() {
+		log.Debug().Msg("closing stream, token expired")
+		cancel()
```

Comment:
will these be cleaned up if the stream breaks?

else I can see the GC holdon to this function, which holds on to the cancel() functions, which holds on to the context

might cause a memory leak?

it returns something you can cancel this with, we should do that (by defer- ing)

Reply:
Instead use context.WithTimeout()